### PR TITLE
macs genome size updated and new logic to calculate it is not present

### LINKED ISSUE #280: --macs_gsize option
**Following on from a discussion on Slack with Harshil Patel, I wanted to raise an issue regarding the genome size needed for the --macs_gsize option when running the nf-core/chipseq pipeline (and others).**

I shall be using genomes not available in iGenomes and so needed to calculate this value myself. To check I could do this correctly, I tried to get the same values for human and mouse as reported at: https://github.com/nf-core/chipseq/blob/master/conf/igenomes.config

To perform the calculation, I ran the script unique-kmers.py as described at:
https://deeptools.readthedocs.io/en/develop/content/feature/effectiveGenomeSize.html

My results for human38 (assuming k-mers of 100-bp) were similar to that reported in iGenomes: 2.8e9 vs 2.7e9 respectively.
However, the calculations for mouse38 were substantially different: 2.47e9 (my calculation) vs 1.87e9 (iGenomes).
(As might be expected, my calculations agree with those displayed on https://deeptools.readthedocs.io/en/develop/content/feature/effectiveGenomeSize.html for 100-bp kmers.)

I spoke to the MACS developers and they appear to agree that these values need updating:
https://github.com/macs3-project/MACS/issues/508#issuecomment-1131448925

**I believe the relevant nf-core documentation should be updated to show these new values.** 

(It may be of interest to you to know that I have been putting together a script that automates the reference genome downloading.  I shall incorporate the DeepTools kmer estimation of genome size into the automated download process.  I shall share this with you when it is ready, incase it is of any use).

Many thanks,

Steven

### PR DESCRIPTION
**Pull request by JoseEspinosa**
Following the discussion on #280 this PR and the response to this [issue](https://github.com/macs3-project/MACS/issues/508), this PR:
- Updates macs genome size values in the `genomes. config` as a hash containing the genome size corresponding to the following read_lengths: 50, 75,100,150 and 200. These values have been calculated as explained [here](https://deeptools.readthedocs.io/en/develop/content/feature/effectiveGenomeSize.html), as shown in the command below for a `read_length` of 50.
```console
unique-kmers.py -k 50 genome.fa
```
- Since the genome size depends on the read length a new parameter `--read_length` has been added that defaults to `null` and gives and gives an error if the pipelines is run without either providing it or the `macs_gsize` param. 
- When the `macs_gsize` value is not present in the `igenomes.config` or the user wants to overwrite it, it is still possible to provide it with `--macs_gsize` parameter.
- If any of the two options above are available, then a new local process `KHMER_UNIQUEKMERS` calculates the macs genome size and provides it to the MACS2_CALLPEAK process.

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Make sure your code lints (`nf-core lint .`).
- [ ] Documentation in `docs` is updated
- [x] `CHANGELOG.md` is updated
- [ ] `README.md` is updated

Closes #280

### CODE REVIEW

File: nextflow_schema.json [hunk 1]
```diff
@@ -33,6 +33,13 @@
                     "description": "Sequencing center information to be added to read group of BAM files.",
                     "fa_icon": "fas fa-synagogue"
                 },
+                "read_length": {
+                    "type": "integer",
+                    "description": "Read length to calculate macs genome size (--macs_gsize)parameter",
+                    "fa_icon": "fas fa-chart-area",
+                    "default": 100,
```
Comment (Emiller88) [inline nextflow_schema.json:40]: This isn't actually the default, the default is `null` currently. So it failed out of the box for me.

**Review by Emiller88 — CHANGES_REQUESTED** [review]
Added a comment about how the docs say the default is `100` but in the `nextflow.config` the value is set to `null`.

Other than that, the code looks excellent!

Another thought, could something like [seqkit stats](https://bioinf.shenwei.me/seqkit/usage/#stats) be used to get the read_length per sample and use that? I don't know if the goal read length or the actual average is better here or not.

**Comment by JoseEspinosa** [comment]
> Added a comment about how the docs say the default is `100` but in the `nextflow.config` the value is set to `null`.
> 
> Other than that, the code looks excellent!
> 
> Another thought, could something like [seqkit stats](https://bioinf.shenwei.me/seqkit/usage/#stats) be used to get the read_length per sample and use that? I don't know if the goal read length or the actual average is better here or not.

We also thought about it, but in the end the idea was discarded since in some cases the read length might not be uniform across samples. See discussion on slack [here](https://nfcore.slack.com/archives/CFP55NZ5G/p1656445226200299?thread_ts=1656433450.126349&cid=CFP55NZ5G). But possibly, we might come with a clever way of doing this for a subsequent release.

File: conf/modules.config [hunk 2]
```diff
@@ -503,16 +503,55 @@ if (!params.skip_plot_fingerprint) {
...
-            ext.args   = [
-                '--keep-dup all',
-                params.narrow_peak      ? '' : "--broad --broad-cutoff ${params.broad_cutoff}",
-                params.save_macs_pileup ? '--bdg --SPMR' : '',
-                params.macs_fdr         ? "--qvalue ${params.macs_fdr}" : '',
-                params.macs_pvalue      ? "--pvalue ${params.macs_pvalue}" : ''
-            ].join(' ').trim()
+        withName: 'HOMER_ANNOTATEPEAKS_MACS2' {
+            ext.args   = '-gid'
+            // ext.prefix = 'peaks'
```
Comment (drpatelh) [inline conf/modules.config:554]: Comment needs to be dealt with here.

File: lib/WorkflowChipseq.groovy [hunk 3]
```diff
@@ -28,6 +28,12 @@ class WorkflowChipseq {
             macsGsizeWarn(log)
         }
 
+        if (!params.read_length && !params.macs_gsize) {
+            log.error "Read length not specified! The pipeline requires either the '--read_length' parameter to be set" +
+            "to infer MACS2 genome size\nor to provide it with the '--macs_gsize' parameter"
```
Comment (drpatelh) [inline lib/WorkflowChipseq.groovy:33]:
```suggestion
            log.error "Both '--read_length' and '--macs_gsize' not specified! Please specify either to infer MACS2 genome size for peak calling."
```

File: lib/WorkflowChipseq.groovy [hunk 4]
```diff
@@ -93,8 +99,8 @@ class WorkflowChipseq {
     private static void macsGsizeWarn(log) {
         log.warn "=============================================================================\n" +
             "  --macs_gsize parameter has not been provided.\n" +
-            "  MACS2 peak-calling and differential analysis will be skipped.\n" +
-            "  Provide '--macs_gsize macs2_genome_size' to change this behaviour.\n" +
+            "  Its value will be obtained using khmer unique-kmers.py based on the --read_length parameter.\n" +
```
Comment (drpatelh) [inline lib/WorkflowChipseq.groovy:102]:
```suggestion
            "  It will be auto-calculated by 'khmer unique-kmers.py' using the '--read_length' parameter.\n" +
```

File: nextflow.config [hunk 5]
```diff
@@ -40,7 +41,6 @@ params {
     // Options: Peaks
     narrow_peak                = false
     broad_cutoff               = 0.1
-    macs_gsize                 = null
```
Comment (drpatelh) [inline nextflow.config:43]: Why was this removed? It is still a valid parameter.

File: nextflow_schema.json [hunk 6]
```diff
@@ -33,6 +33,12 @@
                     "description": "Sequencing center information to be added to read group of BAM files.",
                     "fa_icon": "fas fa-synagogue"
                 },
+                "read_length": {
+                    "type": "integer",
+                    "description": "Read length to calculate macs genome size (--macs_gsize)parameter",
```
Comment (drpatelh) [inline nextflow_schema.json:38]:
```suggestion
                    "description": "Read length use to calculate MACS2 genome size for peak calling if `--macs_gsize` isn't provided.",
```

**Review by drpatelh — COMMENTED** [review]
Looks great @JoseEspinosa ! Minor comments to address.

File: nextflow_schema.json [hunk 7]
```diff
@@ -33,6 +33,12 @@
                     "description": "Sequencing center information to be added to read group of BAM files.",
                     "fa_icon": "fas fa-synagogue"
                 },
+                "read_length": {
+                    "type": "integer",
+                    "description": "Read length use to calculate MACS2 genome size for peak calling if `--macs_gsize` isn't provided.",
```
Comment (drpatelh) [inline nextflow_schema.json:38]:
```suggestion
                    "description": "Read length used to calculate MACS2 genome size for peak calling if `--macs_gsize` isn't provided.",
```

File: conf/modules.config (hunk 2 above)
Comment (drpatelh) [reply conf/modules.config:554]: Just this to sort out now?

Comment (JoseEspinosa) [reply conf/modules.config]: Fix here [9f783fb](https://github.com/nf-core/chipseq/pull/283/commits/9f783fba987971a4002083f0ba0c7341df83e4ee)

**Comment by JoseEspinosa** [comment]
> Looks great @JoseEspinosa ! Minor comments to address.

I think I have addressed all your comments @drpatelh and @Emiller88, would be great if I can get your approval!

**Review by drpatelh — APPROVED** [review]
🚀
